common/workunit/workunit.cpp

```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+    if (clearStats && !wasCreated)
+        sgScopeCollection->clearStats();
+    // Marking the collection dirty here is not ideal.  It would be better to have a call to IStatisticCollection::setStatistic mark the scope as dirty.
+    // However, this would be inefficient as each call to IStatisticCollection::setStatistic would require the dirty flag to be set for all parent scopes.
+    sgScopeCollection->markDirty();
+    return createRootStatisticCollection(creatorType, creator, wfScopeId, graphScopeId, sgScopeCollection);
+}
+
+// Recalculate aggregates for global, workflow and graph scopes
+bool GlobalStatisticCollection::refreshAggregates()
+{
+    return statsCollection->refreshAggregates(aggregateKindsMapping);
+}
+
+// Recalculate aggregates and then write the aggregates to global stats (dali)
+void GlobalStatisticCollection::updateAggregates(IWorkUnit *wu)
+{
+    class StatisticsAggregatesWriter : implements IStatisticVisitor
+    {
+        const StatisticsMapping & aggregateKindsMapping;
+        const unsigned numStats;
+        Linked<IWorkUnit> wu;
+    public:
+        StatisticsAggregatesWriter(IWorkUnit * _wu, const StatisticsMapping & _aggregateKindsMapping): wu(_wu), aggregateKindsMapping(_aggregateKindsMapping), numStats(aggregateKindsMapping.numStatistics()) {}
+
+        virtual bool visitScope(const IStatisticCollection & cur)
         {
-            it->next();
+            switch (cur.queryScopeType())
+            {
+            case SSTglobal:
+            case SSTworkflow:
+            case SSTgraph:
+                for (unsigned i=0; i<numStats; ++i)
+                {
+                    StatisticKind kind = aggregateKindsMapping.getKind(i);
+                    stat_type value;
+                    if (cur.getStatistic(kind, value))
+                    {
+                        if (value || includeStatisticIfZero(kind))
+                        {
+                            StringBuffer s;
+                            wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), cur.queryScopeType(), cur.getFullScope(s).str(), kind, nullptr, value, 1, 0, StatsMergeReplace);
+                        }
+                    }
+                }
+                if (cur.queryScopeType()==SSTgraph)
+                    return false;
+                else
+                    return true;
+            default:
+                return false;
+            }
         }
+    };
+    if (refreshAggregates()) // Only serialize if the aggregates have changed
+    {
+        StatisticsAggregatesWriter statsAggregatorWriter(wu, aggregateKindsMapping);
+        statsCollection->visit(statsAggregatorWriter);
     }
 }
 
-void updateSpillSize(IWorkUnit * wu, const char * scope, StatisticScopeType scopeType)
+// Prune all subgraph descendent stats (leaving subgraph stats for future aggregation)
+void GlobalStatisticCollection::pruneSubGraphDescendants(unsigned wfid, const char *graphName, unsigned sgId)
```

Comment:
There was couple of reasons for implementing the functionality like this
1) to avoid having to copy stats between the gatherer and collector, especially when the updates can be frequent (when lots of small graphs execute).
2) to be able cache stats so that in the future stats may be updated at fixed intervals.  At the moment when many small graphs executed, the stats (GraphProgress) much more frequently than the fixed heartbeat interval defined in the demonserver.  (Updates to any number of sg to GraphProgress may be made in one go.)

I can leave the existing gatherer code as is and merge the necessary stats from the gatherer into the GlobalAggregator.  Let me know if you think we should do this.

Reply:
1. Copying the stats is trivial.  You are copying ~4 stats from one structure to another.
2. That can be solved in a different way.  Saving a queue of blobs to commit.

Let's arrange a walkthrough to discuss before heading off an making any changes though.